### ISSUE TITLE
REST-API basic-auth not stored

### ISSUE THREAD

**Issue by bd8392** [vbier/habpanelviewer#61]
**Before reporting bugs, make sure you have searched for your problem in the openHAB community thread at: https://community.openhab.org/t/habpanelviewer-0-9-25/34112**

**Describe the bug**
If the Rest-Api is behind a basic auth, the auth is not stored. This leads to a problem when reporting valuse, as the habpanelviewer produces logouts through auth-failures by fail2ban. When you restart the app, it asks for the auth of rest-api

**To Reproduce**
1. install openhab behind traefik-reversproxy + fail2ban
2. open the app, define connection, login with basic-auth twice (for web and rest-api)
3. define a reporting sensor item
4. end the app / screenlock
5. get locked out because of auth-faikures

**Expected behavior**
store basic-auth for restapi also

**Screenshots**

**Smartphone/Tablet (please complete the following information):**
 Oneplus 6t Android 10

**Comment by vbier** [comment]
Which version do you use? 0.9.26?

**Comment by bd8392** [comment]
Yes :-) sorry forgot to mention...the latest on F-Droid...

**Comment by bd8392** [comment]
Any news on this?

**Comment by vbier** [comment]
This sounds like requests are blocked by fail2ban, which makes HPV think that the credentials are wrong.

**Comment by bd8392** [comment]
Yes, but I'm only blocked because the credentials are not stored...if I turn of fail2ban, than I get asked again about the credentials...if I type them, I can enter...but I've to do it all the time...not the case for the normal access, but for the rest-api

**Comment by bd8392** [comment]
any news on this? Is there something I can do to help?

**Comment by bd8392** [comment]
I investigated the issue a little bit :-). I turned off fail2ban and tested again. Sometimes - not always - when the screen get's locked and deactivated, habpanelviewer asks for the password of the rest-api after reactivation. If you don't provide it, the app functions, but gets a red border and doesn't report stats anymore...

**Comment by bd8392** [comment]
Any news on this?

**Comment by vbier** [comment]
The problem is that I neither have a password protected rest api nor a fail2ban installation. As you seem to be only one using HPV with fail2ban, the priority is not very high for me. I am going to set up password protected rest api when I have the time and test if it works. If it does not, I will try to fix it. But my time is limited and I have other things to work on so this might take some time.

**Comment by vbier** [comment]
I have just finished adding credential storage encryption to HPV and also tested credential storage. I have uploaded a version you can test: https://github.com/vbier/habpanelviewer/raw/encryptRoom/app-debug.apk

I have changed the code so that the app now memorizes if it has been connected to a password protected server and adds credentials to the rest requests on subsequent starts. So if you start for the first time, it will
- try to connect to your servers rest api without credentials
- ask for the password when it receives a 401 reply and retry the request
- try to load the server web page without credentials
- retry with the credential from the sse connection when it gets a 401 reply

On subsequent starts it will
- try to connect to your servers rest api with the stored credentials
- try to load the server web page without credentials (as there is no way to authorize the request beforehand)
- retry with the credential stored from the sse connection when it gets a 401 reply

I hope that reducing the unauthorized requests will help with fail2ban.

Please try if this helps and report back. Also if you find bugs regarding credential storage.

**Comment by bd8392** [comment]
Wow amazing! Thanks for all the effort and investing time in this! :-) I will test it as soon as possible :-) storing the Rest-Api password will help a lot in securing OpenHab as you now have not limitations with your app :-) great great thanks :-)

**Comment by vbier** [comment]
So did you test it? Does it work?

**Comment by bd8392** [comment]
> So did you test it? Does it work?

No I'm so sorry I completely forgot about it... :-( I will be at my machine in the end of February I then readable Basic Auth (disabled because of upgrade to oh3) and will test it :-). Meanwhile I close this issue because I think it will work :-) great thanks again for caring about it! Love your app very much! I write myself a reminder to report back :-)

**Comment by NoName1311** [comment]
Hi vbier,
because I have the same problem as bd8392, I tested the version provided (https://github.com/vbier/habpanelviewer/raw/encryptRoom/app-debug.apk). The authentication part works fine, but sending commands no longer works.
I'm still on openHab 2.6. The device runs on Android 6.0.1. Is this issue related to my openHab version?
Kind regards

**Comment by vbier** [comment]
Yes, the APK only works with OH3.

**Comment by vbier** [comment]
@NoName1311 , I have just uploaded a new apk to the master branch that contains the fixes I implemented for this issue as well as a new preference entry allowing to select the OH version (right next to the server url). You can try if this works for you, but you have to change the OH version in the preferences to OH2, as otherwise you won't get any item updates.

**Comment by NoName1311** [comment]
Hi @vbier,
thank you for your quick reply and the great app. The new option works perfectly:)
I thought I have to keep my panel project on hold until I move on to 3.0, but with the new version I can follow my plan to update as soon as 3.1 is released. Many thanks.

**Comment by bf8392** [comment]
Confirmed :-) New option is working! Great great Thanks! :-)
